_posts/2020-02-26-demo-fraud-detection-2.md

```diff
@@ -0,0 +1,192 @@
+        .process(new DynamicAlertFunction())
+```
+
+`DynamicKeyFunction`, as described earlier, provides dynamic data partitioning, while `DynamicAlertFunction` is responsible for executing the main logic of processing transactions and sending alert messages according to defined rules.
+
+For a simplified description, we previously assumed that the applied set of rules are pre-initialized and are accessible via the `List<Rules>` within DynamicKeyFunction.
+
+```java
+public class DynamicKeyFunction
+    extends ProcessFunction<Transaction, Keyed<Transaction, String, Integer>> {
+
+  /* Simplified */
+  List<Rule> rules = /* Rules that are initialized somehow.*/;
+  ...
+}
+```
+
+Obviously, adding rules to this list is possible directly inside the code of the Flink Job at the stage of its initialization (create a List object, use it's `add` method). A major drawback of doing so is that it will require recompilation of the job with each rule modification. In the real Fraud Detection system, you can expect to modify these rules on a frequent basis, which makes this approach unacceptable from the point of view of business and operational requirements - a different approach is needed.
+
+Let's next have a look at the example of DSL of one of the sample rules presented in the previous post:
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/rule-dsl.png" width="800px" alt="Figure 1: Rules DSL"/>
+<br/>
+<i><small>Figure 1: Rules DSL</small></i>
+</center>
+<br/>
+![](assets/rule-dsl.png)
+
+Earlier we have already seen how part of this definition, namely `groupingKeyNames`, is used in `DynamicKeyFunction` to extract message keys. At the same time, all other parameters of this rule are necessary within `DynamicAlertFunction` - they define the actual logic of performed operations and it's parameters (such as the alert triggering limit). This means that the same rule must be present in both `DynamicKeyFunction` and `DynamicAlertFunction`. To achieve this result, we will use a special data distribution mechanism in Apache Flink - [events broadcast](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/state/broadcast_state.html). Let’s turn to the Job Graph of the system that we are building:
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/job-graph.png" width="800px" alt="Figure 2: Job Graph"/>
+<br/>
+<i><small>Figure 2: Job Graph</small></i>
+</center>
+<br/>
+![](assets/job-graph.png)
+
+On it you can see the main blocks of our Transactions processing pipeline:<br>
+
+* **Transaction Source**, consuming transaction messages from Kafka partitions in parallel <br>
+    * The forward operator following it means that all data consumed by one of the parallel instances of Transaction Source operator will be transferred to exactly one instance of the subsequent DynamicKeyFunction operator, without redistribution. It also indicates the same level of parallelism of two connected operators (12 in this case).
```

Comment:
Yes, I was not not sure what to call it. I guess reformulating it according to your suggestion hides the fact that we do not have a noun for it :) 
I am, however, not sure about the task/operator terminology when we are talking about the `DynamicKeyFunction`. I was orienting myself on the new Glossary:

Task - 
https://ci.apache.org/projects/flink/flink-docs-stable/concepts/glossary.html#task

Operator - 
https://ci.apache.org/projects/flink/flink-docs-stable/concepts/glossary.html#operator

The way I understand it is that one "instance" of an Operator is a Task, so I would probably stick to my original version here.

@NicoK

Reply:
Your operators are not chained and thus deployed in separate tasks. Communication between tasks is via the network stack via network buffers (in case of forwards, these will thus be exchanged inside a single JVM) while communication between (chained) operators is done directly by calling the next operator with a copy or reference of the output from the current one.
You may see chaining as a special case of forwarding or a fourth communication pattern.
Either way, your job has some chained operators where you do not explain the message passing. The "blocks" you are talking about are tasks.

Reply:
In terms of operator vs. task; an operator is a node of the logical graph and a task is a node of the physical graph. So, technically, there are no communications between operators. You could say "operator instances" everywhere though but we still ignore one important communication pattern: the chain. If you explain communication patterns between operator instances, I think, this one should be in there. Still, I'd rather see tasks and not go into chaining for this blog post (or maybe remove the "normal" communication patterns for good and maybe describe them in the docs instead if there is nothing yet)